src/rust/engine/src/context.rs

```diff
@@ -667,6 +674,91 @@ impl Context {
     )
   }
+
+  ///
+  /// If the given Result is a Failure::MissingDigest, attempts to invalidate the Node which was
+  /// the source of the Digest, potentially causing indirect retry of the Result.
+  ///
+  /// If we successfully locate and restart the source of the Digest, converts the Result into a
+  /// `Failure::Invalidated`, which will cause retry at some level above us.
+  ///
+  pub fn maybe_backtrack(
+    &self,
+    result: NodeResult<NodeOutput>,
+    workunit: &mut RunningWorkunit,
+  ) -> NodeResult<NodeOutput> {
+    let digest = if let Err(Failure::MissingDigest(_, d)) = result.as_ref() {
+      *d
+    } else {
+      return result;
+    };
+
+    // Locate the source(s) of this Digest.
+    // TODO: Currently needs a combination of `visit_live` and `invalidate_from_roots` because
+    // `invalidate_from_roots` cannot view `Node` results. This could lead to a race condition
+    // where a `Node` is invalidated multiple times, which might cause it to increment its attempt
+    // count multiple times.
+    let mut roots = HashSet::new();
+    self.core.graph.visit_live(self, |k, v| match k {
+      NodeKey::ExecuteProcess(p) if v.digests().contains(&digest) => {
+        roots.insert(p.clone());
+      }
+      _ => (),
+    });
+
+    if roots.is_empty() {
+      // We did not identify any roots to invalidate: allow the Node to fail.
+      return result;
+    }
+
+    // Trigger backtrack attempts for the matched Nodes.
+    {
+      let mut backtrack_attempts = self.backtrack_attempts.lock();
+      for root in &roots {
+        let attempt = backtrack_attempts.entry((**root).clone()).or_insert(1);
+        let description = &root.process.description;
+        workunit.increment_counter(Metric::BacktrackAttempts, 1);
+        log::warn!(
+          "Making attempt {attempt} to backtrack and retry `{description}`, due to \
+            missing digest {digest:?}."
+        );
+      }
+    }
+
+    // Invalidate the matched roots.
+    self
+      .core
+      .graph
+      .invalidate_from_roots(move |node| match node {
+        NodeKey::ExecuteProcess(p) => roots.contains(p),
+        _ => false,
+      });
+
+    // We invalidated a Node, and the caller (at some level above us in the stack) should retry.
+    // Complete this node with the Invalidated state.
+    // TODO: Differentiate the reasons for Invalidation (filesystem changes vs missing digests) to
+    // improve warning messages.
+    Err(Failure::Invalidated)
+  }
+
+  ///
+  /// Called before executing a process to determine whether it is backtracking, and if so, to
+  /// increment the attempt count.
+  ///
+  /// A process which has not been marked backtracking will always return 0, regardless of the
+  /// number of calls to this method.
+  ///
+  pub fn maybe_start_backtracking(&self, node: &ExecuteProcess) -> usize {
+    let mut backtrack_attempts = self.backtrack_attempts.lock();
+    let entry: Option<&mut usize> = backtrack_attempts.get_mut(node);
```

Comment:
Rust could not infer type ascription?

Reply:
This is defensive, and partially superstition: if the field is dereferenced too early, the increment can be lost (because you are incrementing a local variable rather a value behind a pointer). The compiler still won't catch that IIRC.